cpp/src/generators/generator_tools.cu

```diff
@@ -0,0 +1,278 @@
+  for (size_t i = 0; i < sources.size(); ++i) size += sources[i].size();
+
+  rmm::device_uvector<vertex_t> srcs_v(size, handle.get_stream());
+  rmm::device_uvector<vertex_t> dsts_v(size, handle.get_stream());
+  rmm::device_uvector<weight_t> weights_v(0, handle.get_stream());
+
+  detail::append_all<vertex_t>(handle, std::move(sources), srcs_v);
+  detail::append_all<vertex_t>(handle, std::move(dests), dsts_v);
+
+  if (optional_d_weights) {
+    weights_v.resize(size, handle.get_stream());
+    detail::append_all(handle, std::move(optional_d_weights.value()), weights_v);
+  }
+
+  size_t number_of_edges{0};
+
+  if (optional_d_weights) {
+    thrust::sort(
+      rmm::exec_policy(handle.get_stream()),
+      thrust::make_zip_iterator(
+        thrust::make_tuple(srcs_v.begin(), dsts_v.begin(), weights_v.begin())),
+      thrust::make_zip_iterator(thrust::make_tuple(srcs_v.end(), dsts_v.end(), weights_v.end())));
+
+    auto pair_first = thrust::make_zip_iterator(thrust::make_tuple(srcs_v.begin(), dsts_v.begin()));
+    auto end_iter   = thrust::unique_by_key(rmm::exec_policy(handle.get_stream()),
+                                          pair_first,
+                                          pair_first + srcs_v.size(),
+                                          weights_v.begin());
+
+    number_of_edges = thrust::distance(pair_first, thrust::get<0>(end_iter));
+  } else {
+    thrust::sort(rmm::exec_policy(handle.get_stream()),
+                 thrust::make_zip_iterator(thrust::make_tuple(srcs_v.begin(), dsts_v.begin())),
+                 thrust::make_zip_iterator(thrust::make_tuple(srcs_v.end(), dsts_v.end())));
+
+    auto pair_first = thrust::make_zip_iterator(thrust::make_tuple(srcs_v.begin(), dsts_v.begin()));
+
+    auto end_iter =
+      thrust::unique(rmm::exec_policy(handle.get_stream()),
+                     thrust::make_zip_iterator(thrust::make_tuple(srcs_v.begin(), dsts_v.begin())),
+                     thrust::make_zip_iterator(thrust::make_tuple(srcs_v.end(), dsts_v.end())));
+
+    number_of_edges = thrust::distance(pair_first, end_iter);
+  }
+
+  srcs_v.resize(number_of_edges, handle.get_stream());
+  srcs_v.shrink_to_fit(handle.get_stream());
+  dsts_v.resize(number_of_edges, handle.get_stream());
+  dsts_v.shrink_to_fit(handle.get_stream());
+
+  if (optional_d_weights) {
+    weights_v.resize(number_of_edges, handle.get_stream());
+    weights_v.shrink_to_fit(handle.get_stream());
+  }
+
+  return std::make_tuple(std::move(srcs_v), std::move(dsts_v), std::move(weights_v));
+}
+
+template <typename vertex_t, typename weight_t>
+std::
+  tuple<rmm::device_uvector<vertex_t>, rmm::device_uvector<vertex_t>, rmm::device_uvector<weight_t>>
+  symmetrize_edgelist(raft::handle_t const &handle,
+                      rmm::device_uvector<vertex_t> &&d_src_v,
+                      rmm::device_uvector<vertex_t> &&d_dst_v,
+                      std::optional<rmm::device_uvector<weight_t>> &&optional_d_weights_v)
+{
```

Comment:
I think the code below is simpler and more efficient (in memory footprint & the amount of data movement).
```
auto offset = d_src_v.size();
d_src_v.resize(offset * 2, handle.get_stream_view());
d_dst_v.resize(d_src_v.size(), handle.get_stream_view());
thrust::copy(rmm::exec_policy(handle.get_stream_view()), d_dst_v.begin(), d_dst_v.begin() + offset, d_src_v.begin() + offset);
thrust::copy(rmm::exec_policy(handle.get_stream_view()), d_src_v.begin(), d_src_v.begin() + offset, d_dst_v.begin() + offset);
if (optional_d_weights_v) {
  optional_d_weights_v.resize(d_src_v.size(), handle.get_stream_view());
  thrust::copy(rmm::exec_policy(handle.get_stream_view()), optional_d_weight_v.begin(), optinoal_d_weight_v.begin() + offset, optional_d_weight_v.begin() + offset);
}

return std::make_tuple(std::move(d_src_v), std::move(d_dst_v), optional_d_weights_v ? std::move(optional_d_weights_v) : std::nullopt);
```

Reply:
Implemented this version.